### ISSUE TITLE
Citus might fail to assign correct name to index on partition shard when index name is long

### ISSUE THREAD

**Issue by onurctirtir** [citusdata/citus#4962]
```sql
SELECT citus_add_node('localhost', 9701);

CREATE TABLE part_table_with_very_long_name (
    dist_col integer,
    long_named_integer_col integer,
    long_named_part_col timestamp
) PARTITION BY RANGE (long_named_part_col);

CREATE TABLE part_table_with_long_long_long_long_name
PARTITION OF part_table_with_very_long_name
FOR VALUES FROM ('2010-01-01') TO ('2015-01-01');

SELECT create_distributed_table('part_table_with_very_long_name', 'dist_col');

CREATE INDEX ON part_table_with_very_long_name
USING btree (long_named_integer_col, long_named_part_col);
```

When creating index for shell table during metadata sync, we get this error:
```sql
SELECT start_metadata_sync_to_node ('localhost', 9701);
ERROR:  relation "part_table_with_long_long_lon_long_named_integer_col_long_n_idx" already exists
CONTEXT:  while executing command on localhost:9701
```

This is because, we somehow already assigned that index name to one of the shards:
```sql
select tablename from pg_indexes where indexname = 'part_table_with_long_long_lon_long_named_integer_col_long_n_idx';
┌─────────────────────────────────────────────────┐
│                    tablename                    │
├─────────────────────────────────────────────────┤
│ part_table_with_long_long_long_long_name_102040 │
└─────────────────────────────────────────────────┘
(1 row)
```

**Comment by onderkalaci** [comment]
> This is because, we somehow already assigned that index name to one of the shards:

Citus doesn't explicitly assigns this name, but when an index is created on the parent table, Postgres auto-generates the index name on the children.

```SQL
CREATE TABLE parent (
    dist_col integer,
    long_named_integer_col integer,
    long_named_part_col timestamp
) PARTITION BY RANGE (long_named_part_col);


CREATE TABLE child
PARTITION OF parent
FOR VALUES FROM ('2010-01-01') TO ('2015-01-01');

CREATE INDEX ON parent
USING btree (long_named_integer_col, long_named_part_col);


\d child
                                 Table "public.child"
         Column         |            Type             | Collation | Nullable | Default 
------------------------+-----------------------------+-----------+----------+---------
 dist_col               | integer                     |           |          | 
 long_named_integer_col | integer                     |           |          | 
 long_named_part_col    | timestamp without time zone |           |          | 
Partition of: parent FOR VALUES FROM ('2010-01-01 00:00:00') TO ('2015-01-01 00:00:00')
Indexes:
    "child_long_named_integer_col_long_named_part_col_idx" btree (long_named_integer_col, long_named_part_col)
```

And, for the Citus case, the auto generated index names of the shell table on the coordinator and the first shard on the workers becomes the same.

@onurctirtir does this only happen with long table names?   For reference: https://github.com/postgres/postgres/blob/197d33ccbe888fc84ae4e49bb241e88ea3c81f15/src/backend/commands/indexcmds.c#L2329

**Comment by onderkalaci** [comment]
before https://github.com/citusdata/citus/pull/4273 we get:
```SQL
ERROR:  creating index without a name on a distributed table is currently unsupported
Time: 0.948 ms
```

So, this is a bug introduced on Citus 10, we should probably backport

**Comment by onurctirtir** [comment]
>@onurctirtir does this only happen with long table names? For reference: https://github.com/postgres/postgres/blob/197d33ccbe888fc84ae4e49bb241e88ea3c81f15/src/backend/commands/indexcmds.c#L2329

I couldn't reproduce with shorter index names, let me take another look on Monday.

**Comment by onderkalaci** [comment]
`This would have been prevented if `CREATE INDEX` statement has explicitly stated the index name like `CREATE INDEX index_name ON part_table_with_very_long_name
USING btree (long_named_integer_col, long_named_part_col);`

**Comment by onurctirtir** [comment]
I get the same error if I use the same long index name that postgres would assign, even on v9.5.4:

```diff
SELECT master_add_node('localhost', 10701);

CREATE TABLE part_table_with_very_long_name (
    dist_col integer,
    long_named_integer_col integer,
    long_named_part_col timestamp
) PARTITION BY RANGE (long_named_part_col);

CREATE TABLE part_table_with_long_long_long_long_name
PARTITION OF part_table_with_very_long_name
FOR VALUES FROM ('2010-01-01') TO ('2015-01-01');

SELECT create_distributed_table('part_table_with_very_long_name', 'dist_col');

-CREATE INDEX
+CREATE INDEX part_table_with_very_long_nam_long_named_integer_col_long_n_idx
ON part_table_with_very_long_name
USING btree (long_named_integer_col, long_named_part_col);

SELECT start_metadata_sync_to_node ('localhost', 10701);
ERROR:  relation "part_table_with_long_long_lon_long_named_integer_col_long_n_idx" already exists
CONTEXT:  while executing command on localhost:10701
```

**Comment by onurctirtir** [comment]
Moreover, we don't need give a long name to parent or to it's index.

Index name that postgres would chose for child table should be long enough, either by having a long name for child table or by having long column names in parent relation:

```sql
SELECT master_add_node('localhost', 10701);

CREATE TABLE parent (
    dist_col integer,
    int_col integer,
    part_col timestamp
) PARTITION BY RANGE (part_col);

CREATE TABLE child_child_child_child_child_child_child_child_child
PARTITION OF parent
FOR VALUES FROM ('2010-01-01') TO ('2015-01-01');

SET citus.replication_model TO streaming;

SELECT create_distributed_table('parent', 'dist_col');

CREATE INDEX a_short_idx_name
ON parent
USING btree (int_col, part_col);

SELECT start_metadata_sync_to_node ('localhost', 10701);
ERROR:  relation "part_table_with_long_long_lon_int_col_long_n_idx" already exists
CONTEXT:  while executing command on localhost:10701
```

**Comment by onurctirtir** [comment]
The [solution](https://github.com/citusdata/citus/issues/4962#issuecomment-838371718) I shared yesterday was obviously too complex, so with @onderkalaci, we came up another way to fix the issue:

Intercept `CREATE INDEX` commands. If command is run for a parent relation then:
1) Run given `CREATE INDEX` by forcing `ONLY` behavior.
2) Create the very same index on all partitions (so that citus assigns correct names to indexes on partition shards)
3) Attach indexes that we created on partitions to parent table index

For 1), we need to fix #4938 and for 3), we need to fix #4938.

Also, in addition to intercepting such `CREATE INDEX` commands, we should also handle `CREATE TABLE PARTITION OF` commands. This is because, such commands do also create the indexes inherited from parent table on new partition.

Other than that, we still need to implement something like the [solution](https://github.com/citusdata/citus/issues/4962#issuecomment-838371718) that I shared yesterday for upgrade script, to fix such existing & broken partition index names during citus upgrades.

**Comment by naisila** [comment]
Note:
The following scenario will have two index-name clashes:
```SQL
CREATE TABLE part_table_with_very_long_name (
    dist_col integer,
    long_named_integer_col integer,
    long_named_part_col timestamp
) PARTITION BY RANGE (long_named_part_col);

CREATE TABLE part_table_with_long_long_long_long_name
PARTITION OF part_table_with_very_long_name
FOR VALUES FROM ('2010-01-01') TO ('2015-01-01');

CREATE TABLE part_table_with_long_long_long_long_name_2
PARTITION OF part_table_with_very_long_name
FOR VALUES FROM ('2016-01-01') TO ('2017-01-01');

SELECT create_distributed_table('part_table_with_very_long_name', 'dist_col');

CREATE INDEX ON part_table_with_very_long_name
USING btree (long_named_integer_col, long_named_part_col);
```

```SQL
SELECT start_metadata_sync_to_node ('localhost', 9701);
ERROR:  relation "part_table_with_long_long_lon_long_named_integer_col_long_n_idx" already exists
CONTEXT:  while executing command on localhost:9701
```
`part_table_with_long_long_lon_long_named_integer_col_long_n_idx` is the index name for first partition. If we fix that one, we will still get the following error because `part_table_with_long_long_lon_long_named_integer_col_long__idx1` is the index name of the second partition. The list of errors may go on since all shards of partitions will have names ending in `idx1, idx2, ... idxn` and these names do not include the shardid, and could very well be the name of the index of a shell table.
```SQL
-- assuming we fixed the previous error with part_table_with_long_long_lon_long_named_integer_col_long_n_idx
SELECT start_metadata_sync_to_node ('localhost', 9701);
ERROR:  relation "part_table_with_long_long_lon_long_named_integer_col_long__idx1" already exists
CONTEXT:  while executing command on localhost:9701
```
